File: BubblerApp/BubblerApp/Components/PreferenceTopicsEditor.swift
//
//  PreferenceTopicsEditor.swift
//  BubblerApp
//

import SwiftUI

struct PreferenceTopicsEditor: View {
    let title: String
    let subtitle: String
    let icon: String
    let iconColor: Color
    @Binding var topics: [String]

    private let conflictingTopics: Binding<[String]>?

    @State private var draft = ""
    @State private var errorMessage: String?

    init(
        title: String,
        subtitle: String,
        icon: String,
        iconColor: Color,
        topics: Binding<[String]>,
        conflictingTopics: Binding<[String]>? = nil
    ) {
        self.title = title
        self.subtitle = subtitle
        self.icon = icon
        self.iconColor = iconColor
        _topics = topics
        self.conflictingTopics = conflictingTopics
    }

    private var matchingTopics: [String] {
        KnownTopics.matching(draft, excluding: topics)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            VStack(alignment: .leading, spacing: 4) {
                HStack(spacing: 8) {
                    Image(systemName: icon)
                        .foregroundColor(iconColor)

                    Text(title)
                        .font(.headline)
                        .foregroundColor(.white)
                }

                Text(subtitle)
                    .font(.caption)
                    .foregroundColor(.white.opacity(0.65))
            }

            searchField

            if let errorMessage {
                Text(errorMessage)
                    .font(.caption.weight(.semibold))
                    .foregroundColor(.red.opacity(0.95))
                    .frame(maxWidth: .infinity, alignment: .leading)
            }

            if !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                suggestionResults
            }

            if topics.isEmpty {
                Text("No topics added yet.")
                    .font(.caption)
                    .foregroundColor(.white.opacity(0.6))
                    .frame(maxWidth: .infinity, alignment: .leading)
            } else {
                TopicChipGrid(topics: topics, onRemove: removeTopic)
            }
        }
        .padding(18)
        .background(
            RoundedRectangle(cornerRadius: 22)
                .fill(Color.white.opacity(0.08))
                .overlay(
                    RoundedRectangle(cornerRadius: 22)
                        .stroke(Color.white.opacity(0.12), lineWidth: 1)
                )
        )
    }

    private var searchField: some View {
        HStack(spacing: 10) {
            Image(systemName: "magnifyingglass")
                .foregroundColor(iconColor.opacity(0.9))

            TextField("Search existing topics...", text: $draft)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .submitLabel(.done)
                .onSubmit(addTopic)
                .onChange(of: draft) { _, _ in
                    errorMessage = nil
                }
                .foregroundColor(.white)

            if !draft.isEmpty {
                Button {
                    draft = ""
                    errorMessage = nil
                } label: {
                    Image(systemName: "xmark.circle.fill")
                        .foregroundColor(.white.opacity(0.55))
                }
                .accessibilityLabel("Clear topic search")
            }

            Button("Add", action: addTopic)
                .foregroundColor(.white)
                .font(.subheadline.weight(.semibold))
        }
        .padding(12)
        .background(Color.white.opacity(0.08))
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .stroke(Color.white.opacity(0.12), lineWidth: 1)
        )
    }

    @ViewBuilder
    private var suggestionResults: some View {
        if matchingTopics.isEmpty {
            Text("No existing topics match \"\(draft.trimmingCharacters(in: .whitespacesAndNewlines))\".")
                .font(.caption)
                .foregroundColor(.white.opacity(0.6))
                .frame(maxWidth: .infinity, alignment: .leading)
        } else {
            VStack(spacing: 8) {
                ForEach(matchingTopics, id: \.self) { topic in
                    Button {
                        selectTopic(topic)
                    } label: {
                        HStack(spacing: 10) {
                            Image(systemName: TopicStyle.icon(for: topic))
                                .foregroundColor(TopicStyle.color(for: topic))
                                .frame(width: 18)

                            Text(KnownTopics.displayName(for: topic))
                                .font(.subheadline.weight(.semibold))
                                .foregroundColor(.white)

                            Spacer()

                            Image(systemName: "plus.circle.fill")
                                .foregroundColor(.white.opacity(0.55))
                        }
                        .padding(.horizontal, 12)
                        .padding(.vertical, 10)
                        .background(Color.white.opacity(0.1))
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                    }
                    .buttonStyle(.plain)
                }
            }
        }
    }

    private func addTopic() {
        let trimmed = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            errorMessage = nil
            return
        }

        guard let topic = KnownTopics.resolve(trimmed) else {
            errorMessage = "Unknown topic: \"\(trimmed)\". Choose one from the list of existing topics."
            return
        }

        selectTopic(topic)
    }

    private func selectTopic(_ topic: String) {
        guard let resolved = KnownTopics.resolve(topic) else {
            errorMessage = "Unknown topic: \"\(topic)\". Choose one from the list of existing topics."
            return
        }

        // Clear the other list first so preferred → blacklisted doesn't get
        // dropped by merge (preferred wins when both are present).
        if let conflictingTopics {
            conflictingTopics.wrappedValue = TopicPreferenceList.remove(
                resolved,
                from: conflictingTopics.wrappedValue
            )
        }
        topics = TopicPreferenceList.add(resolved, to: topics)
        draft = ""
        errorMessage = nil
    }

    private func removeTopic(_ topic: String) {
        topics = TopicPreferenceList.remove(topic, from: topics)
    }
}

private struct TopicChipGrid: View {
    let topics: [String]
    let onRemove: (String) -> Void

    var body: some View {
        LazyVGrid(columns: [GridItem(.adaptive(minimum: 120), spacing: 10)], alignment: .leading, spacing: 10) {
            ForEach(topics, id: \.self) { topic in
                HStack(spacing: 8) {
                    Text(KnownTopics.displayName(for: topic))
                        .font(.caption.weight(.semibold))
                        .lineLimit(1)

                    Button {
                        onRemove(topic)
                    } label: {
                        Image(systemName: "xmark.circle.fill")
                            .font(.caption)
                    }
                    .buttonStyle(.plain)
                }
                .foregroundColor(.white)
                .padding(.horizontal, 12)
                .padding(.vertical, 9)
                .background(Color.white.opacity(0.1))
                .clipShape(Capsule())
                .overlay(
                    Capsule()
                        .stroke(Color.white.opacity(0.12), lineWidth: 1)
                )
            }
        }
    }
}

File: bubbler_app/lib/shared/widgets/preference_topics_editor.dart
import 'package:flutter/material.dart';

import '../../data/models/topics.dart';
import '../theme/topic_style.dart';

/// Searchable preferred/blacklist topic editor — Swift `PreferenceTopicsEditor`.
class PreferenceTopicsEditor extends StatefulWidget {
  const PreferenceTopicsEditor({
    super.key,
    required this.title,
    required this.subtitle,
    required this.icon,
    required this.iconColor,
    required this.topics,
    required this.onChanged,
    this.conflictingTopics,
    this.onConflictingChanged,
  });

  final String title;
  final String subtitle;
  final IconData icon;
  final Color iconColor;
  final List<String> topics;
  final ValueChanged<List<String>> onChanged;

  /// When set, selecting a topic also removes it from this conflicting list
  /// (preferred ↔ blacklisted), matching iOS merge behavior.
  final List<String>? conflictingTopics;
  final ValueChanged<List<String>>? onConflictingChanged;

  @override
  State<PreferenceTopicsEditor> createState() => _PreferenceTopicsEditorState();
}

class _PreferenceTopicsEditorState extends State<PreferenceTopicsEditor> {
  final _draftController = TextEditingController();
  String? _errorMessage;

  @override
  void dispose() {
    _draftController.dispose();
    super.dispose();
  }

  List<String> get _matchingTopics =>
      KnownTopics.matching(_draftController.text, excluding: widget.topics);

  void _addTopic() {
    final trimmed = _draftController.text.trim();
    if (trimmed.isEmpty) {
      setState(() => _errorMessage = null);
      return;
    }

    final topic = KnownTopics.resolve(trimmed);
    if (topic == null) {
      setState(() {
        _errorMessage =
            'Unknown topic: "$trimmed". Choose one from the list of existing topics.';
      });
      return;
    }

    _selectTopic(topic);
  }

  void _selectTopic(String topic) {
    final resolved = KnownTopics.resolve(topic);
    if (resolved == null) {
      setState(() {
        _errorMessage =
            'Unknown topic: "$topic". Choose one from the list of existing topics.';
      });
      return;
    }

    // Clear the other list first so preferred → blacklisted doesn't get
    // dropped by merge (preferred wins when both are present).
    final conflicting = widget.conflictingTopics;
    final onConflicting = widget.onConflictingChanged;
    if (conflicting != null && onConflicting != null) {
      onConflicting(TopicPreferenceList.remove(resolved, conflicting));
    }

    widget.onChanged(TopicPreferenceList.add(resolved, widget.topics));
    _draftController.clear();
    setState(() => _errorMessage = null);
  }

  void _removeTopic(String topic) {
    widget.onChanged(TopicPreferenceList.remove(topic, widget.topics));
  }

  @override
  Widget build(BuildContext context) {
    final draft = _draftController.text.trim();

    return Container(
      padding: const EdgeInsets.all(18),
      decoration: BoxDecoration(
        color: Colors.white.withValues(alpha: 0.08),
        borderRadius: BorderRadius.circular(22),
        border: Border.all(color: Colors.white.withValues(alpha: 0.12)),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              Icon(widget.icon, color: widget.iconColor, size: 18),
              const SizedBox(width: 8),
              Expanded(
                child: Text(
                  widget.title,
                  style: const TextStyle(
                    color: Colors.white,
                    fontSize: 17,
                    fontWeight: FontWeight.w600,
                  ),
                ),
              ),
            ],
          ),
          const SizedBox(height: 4),
          Text(
            widget.subtitle,
            style: TextStyle(
              color: Colors.white.withValues(alpha: 0.65),
              fontSize: 12,
            ),
          ),
          const SizedBox(height: 14),
          _SearchField(
            controller: _draftController,
            iconColor: widget.iconColor,
            onChanged: (_) => setState(() => _errorMessage = null),
            onSubmitted: (_) => _addTopic(),
            onClear: () {
              _draftController.clear();
              setState(() => _errorMessage = null);
            },
            onAdd: _addTopic,
          ),
          if (_errorMessage != null) ...[
            const SizedBox(height: 10),
            Text(
              _errorMessage!,
              style: TextStyle(
                color: Colors.red.withValues(alpha: 0.95),
                fontSize: 12,
                fontWeight: FontWeight.w600,
              ),
            ),
          ],
          if (draft.isNotEmpty) ...[
            const SizedBox(height: 10),
            _SuggestionResults(
              draft: draft,
              matching: _matchingTopics,
              onSelect: _selectTopic,
            ),
          ],
          const SizedBox(height: 14),
          if (widget.topics.isEmpty)
            Text(
              'No topics added yet.',
              style: TextStyle(
                color: Colors.white.withValues(alpha: 0.6),
                fontSize: 12,
              ),
            )
          else
            _TopicChipGrid(
              topics: widget.topics,
              onRemove: _removeTopic,
            ),
        ],
      ),
    );
  }
}

class _SearchField extends StatelessWidget {
  const _SearchField({
    required this.controller,
    required this.iconColor,
    required this.onChanged,
    required this.onSubmitted,
    required this.onClear,
    required this.onAdd,
  });

  final TextEditingController controller;
  final Color iconColor;
  final ValueChanged<String> onChanged;
  final ValueChanged<String> onSubmitted;
  final VoidCallback onClear;
  final VoidCallback onAdd;

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 4),
      decoration: BoxDecoration(
        color: Colors.white.withValues(alpha: 0.08),
        borderRadius: BorderRadius.circular(14),
        border: Border.all(color: Colors.white.withValues(alpha: 0.12)),
      ),
      child: Row(
        children: [
          Icon(Icons.search, color: iconColor.withValues(alpha: 0.9), size: 18),
          const SizedBox(width: 10),
          Expanded(
            child: TextField(
              controller: controller,
              onChanged: onChanged,
              onSubmitted: onSubmitted,
              autocorrect: false,
              enableSuggestions: false,
              textCapitalization: TextCapitalization.none,
              textInputAction: TextInputAction.done,
              style: const TextStyle(color: Colors.white),
              cursorColor: Colors.white,
              decoration: InputDecoration(
                hintText: 'Search existing topics...',
                hintStyle: TextStyle(
                  color: Colors.white.withValues(alpha: 0.45),
                ),
                border: InputBorder.none,
                isDense: true,
              ),
            ),
          ),
          if (controller.text.isNotEmpty)
            IconButton(
              onPressed: onClear,
              icon: Icon(
                Icons.cancel,
                color: Colors.white.withValues(alpha: 0.55),
                size: 18,
              ),
              tooltip: 'Clear topic search',
              visualDensity: VisualDensity.compact,
            ),
          TextButton(
            onPressed: onAdd,
            child: const Text(
              'Add',
              style: TextStyle(
                color: Colors.white,
                fontWeight: FontWeight.w600,
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class _SuggestionResults extends StatelessWidget {
  const _SuggestionResults({
    required this.draft,
    required this.matching,
    required this.onSelect,
  });

  final String draft;
  final List<String> matching;
  final ValueChanged<String> onSelect;

  @override
  Widget build(BuildContext context) {
    if (matching.isEmpty) {
      return Text(
        'No existing topics match "$draft".',
        style: TextStyle(
          color: Colors.white.withValues(alpha: 0.6),
          fontSize: 12,
        ),
      );
    }

    return Column(
      children: [
        for (final topic in matching) ...[
          Material(
            color: Colors.transparent,
            child: InkWell(
              onTap: () => onSelect(topic),
              borderRadius: BorderRadius.circular(12),
              child: Ink(
                padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 10),
                decoration: BoxDecoration(
                  color: Colors.white.withValues(alpha: 0.1),
                  borderRadius: BorderRadius.circular(12),
                ),
                child: Row(
                  children: [
                    SizedBox(
                      width: 18,
                      child: Icon(
                        TopicStyle.icon(topic),
                        size: 16,
                        color: TopicStyle.color(topic),
                      ),
                    ),
                    const SizedBox(width: 10),
                    Expanded(
                      child: Text(
                        KnownTopics.displayName(topic),
                        style: const TextStyle(
                          color: Colors.white,
                          fontSize: 14,
                          fontWeight: FontWeight.w600,
                        ),
                      ),
                    ),
                    Icon(
                      Icons.add_circle,
                      color: Colors.white.withValues(alpha: 0.55),
                      size: 18,
                    ),
                  ],
                ),
              ),
            ),
          ),
          const SizedBox(height: 8),
        ],
      ],
    );
  }
}

class _TopicChipGrid extends StatelessWidget {
  const _TopicChipGrid({
    required this.topics,
    required this.onRemove,
  });

  final List<String> topics;
  final ValueChanged<String> onRemove;

  @override
  Widget build(BuildContext context) {
    return Wrap(
      spacing: 10,
      runSpacing: 10,
      children: [
        for (final topic in topics)
          Container(
            padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 9),
            decoration: BoxDecoration(
              color: Colors.white.withValues(alpha: 0.1),
              borderRadius: BorderRadius.circular(999),
              border: Border.all(color: Colors.white.withValues(alpha: 0.12)),
            ),
            child: Row(
              mainAxisSize: MainAxisSize.min,
              children: [
                Text(
                  KnownTopics.displayName(topic),
                  style: const TextStyle(
                    color: Colors.white,
                    fontSize: 12,
                    fontWeight: FontWeight.w600,
                  ),
                ),
                const SizedBox(width: 8),
                GestureDetector(
                  onTap: () => onRemove(topic),
                  child: const Icon(Icons.cancel, size: 14, color: Colors.white),
                ),
              ],
            ),
          ),
      ],
    );
  }
}
